ai-platform-unified/tutorials/unofficial/pytorch-image-classification-custom-train-container-gpu-custom-prediction-container-torchserve/README.md

```diff
@@ -0,0 +1,243 @@
+rm -rf model
+```
+
+### Custom PyTorch Container for Training
+
+```
+export HOSTNAME=gcr.io
+export IMAGE_TRAIN=${TUTORIAL_NAME_TRAIN}
+export TAG=latest
+export CONTAINER_IMAGE_URI=${HOSTNAME}/${PROJECT}/${IMAGE_TRAIN}:${TAG}
+
+echo "CONTAINER_IMAGE_URI: ${CONTAINER_IMAGE_URI}"
+```
+```
+cd ${TUTORIAL_DIR}
+cd trainer && docker build -t ${CONTAINER_IMAGE_URI} -f Dockerfile .
+```
+```
+docker run --rm ${CONTAINER_IMAGE_URI} --epochs 1
+```
+```
+docker push ${CONTAINER_IMAGE_URI}
+```
+```
+gcloud container images list --repository ${HOSTNAME}/${PROJECT}
+```
+
+### Launch a Custom Training Job
+
+##### Configs
+###### CPU
+```
+export CUST_TRAIN_DISPLAY_NAME_SDK=${TUTORIAL_NAME_TRAIN}-sdk-cpu
+
+export MACHINE_TYPE=n1-standard-4
+export ACCELERATOR_TYPE=ACCELERATOR_TYPE_UNSPECIFIED
+export ACCELERATOR_COUNT=0
+export CONT_ARGS="--batch-size 4 --epochs 5"
+
+```
+
+###### Single GPU
+```
+export CUST_TRAIN_DISPLAY_NAME_SDK=${TUTORIAL_NAME_TRAIN}-sdk-gpu
+
+export MACHINE_TYPE=n1-standard-4
+export ACCELERATOR_TYPE=NVIDIA_TESLA_K80
+export ACCELERATOR_COUNT=1
+export CONT_ARGS="--batch-size 4 --epochs 5"
+```
+
+###### Multiple GPUs
+```
+export CUST_TRAIN_DISPLAY_NAME_SDK=${TUTORIAL_NAME_TRAIN}-sdk-gpus
+
+export MACHINE_TYPE=n1-standard-32
+export ACCELERATOR_TYPE=NVIDIA_TESLA_P100
+export ACCELERATOR_COUNT=4
+export CONT_ARGS="--batch-size 32 --epochs 5"
+```
+
+##### SDK
+```
+echo "CUST_TRAIN_DISPLAY_NAME_SDK: ${CUST_TRAIN_DISPLAY_NAME_SDK}"
+
+echo "MACHINE_TYPE: ${MACHINE_TYPE}"
+echo "ACCELERATOR_COUNT: ${ACCELERATOR_COUNT}"
+echo "ACCELERATOR_TYPE: ${ACCELERATOR_TYPE}"
+echo "CONT_ARGS: ${CONT_ARGS} --epochs 5"
+
+cd ${TUTORIAL_DIR}
+python sdk_custom_container_training.py \
+    --project ${PROJECT} \
+    --bucket ${BUCKET} \
+    --location ${LOCATION} \
+    --display-name ${CUST_TRAIN_DISPLAY_NAME_SDK} \
+    --container-image-uri ${CONTAINER_IMAGE_URI} \
+    --machine-type ${MACHINE_TYPE} \
+    --accelerator-type ${ACCELERATOR_TYPE} \
+    --accelerator-count ${ACCELERATOR_COUNT} \
+    --container-args ${CONT_ARGS}
+```
+###### Output Model Artifact
+```
+gsutil ls gs://${BUCKET}/${CUST_TRAIN_DISPLAY_NAME_SDK}
+```
+
+
+### Model Archive for TorchServe
+
+```
+cd ${TUTORIAL_DIR}
+cd model_server && gsutil cp -r gs://${BUCKET}/${CUST_TRAIN_DISPLAY_NAME_SDK}/model .
+```
+```
+cd ${TUTORIAL_DIR}
+cd model_server && torch-model-archiver \
+     --model-name antandbee \
+     --version 1.0 \
+     --serialized-file ./model/antandbee.pth \
+     --model-file ./model.py \
+     --handler ./handler.py \
+     --extra-files ./index_to_name.json \
+     -f
+```
+### TorchServe Local Run
+```
+cd ${TUTORIAL_DIR}
+cd model_server && torchserve \
+    --model-store ./ \
+    --ts-config ./config.properties \
+    --models antandbee=antandbee.mar
+```
+
+```
+curl http://localhost:8080/ping
+curl http://127.0.0.1:8081/models/antandbee
+
+wget https://raw.githubusercontent.com/alvarobartt/pytorch-model-serving/master/images/sample.jpg
```

Comment:
Nits:

1. Not sure if there is a best practice around this, but since `wget` is not installed on Macs by default and since you use `curl` elsewhere, I think you could replace this line (and other `wget` lines) with the following for the same effect:

   ```
   curl -O https://raw.githubusercontent.com/alvarobartt/pytorch-model-serving/master/images/sample.jpg
   ```

   This might make the commands more consistent + work on more devices.

2. Again, not sure what the best practice is, and the open-source license for `alvarobartt/pytorch-model-serving` seems fine, but would it be better to host a copy of the sample image yourself either in this GitHub repository or in `gs://cloud-samples-data`? This way this example doesn't break based if the external dependency changes.

(Also, the sample image seems to be of a hamburger, which doesn't seem like the most appropriate for this particular model. 😆  Is there a different one you intended to use?)